crypto/evp/p_seal.c

```diff
@@ -21,16 +21,18 @@ int EVP_SealInit(EVP_CIPHER_CTX *ctx, const EVP_CIPHER *type,
                  EVP_PKEY **pubk, int npubk)
 {
     unsigned char key[EVP_MAX_KEY_LENGTH];
-    const OSSL_PROVIDER *prov = EVP_CIPHER_provider(type);
-    OSSL_LIB_CTX *libctx = prov != NULL ? ossl_provider_libctx(prov) : NULL;
+    const OSSL_PROVIDER *prov = NULL;
+    OSSL_LIB_CTX *libctx = NULL;
     EVP_PKEY_CTX *pctx = NULL;
     int i, len;
     int rv = 0;
 
-    if (type) {
+    if (type != NULL) {
         EVP_CIPHER_CTX_reset(ctx);
         if (!EVP_EncryptInit_ex(ctx, type, NULL, NULL, NULL))
             return 0;
+        if ((prov = EVP_CIPHER_provider(type)) != NULL)
```

Comment:
This should be after the if() and the type should be obtained from ctx->cipher instead.

Reply:
No it shouldn't.  This is a use case (from evp_extra_test):
```C
ctx = EVP_CIPHER_CTX_new();
EVP_SealInit(ctx, type, ...);
```

Reply:
Which makes the fix more complicated.

If (type != NULL) use type's libctx else use ctx's libctx.

Reply:
Update pushed.

Reply:
I think this is needlessly complicated. You can fairly well depend on the `EVP_EncryptInit_ex(ctx, type, NULL, NULL, NULL)` call setting the ctx->cipher == type. At least for a fetched `type`. Which is actually the only kind of `type` that will have non-NULL provider anyway.